src/analysis/output.js

```diff
@@ -0,0 +1,104 @@
+export type Index = Integer;
+
+/**
+ * Describes an individual node in the contribution graph.
+ * Includes the information in the raw Graph node, along with scoring
+ * information, i.e. its total cred, and the amount of cred that it minted.
+ *
+ * While we are in TimelineCred, the "cred" field will have the total cred,
+ * summed across all time slices. Once we migrate to CredRank, it will just be
+ * the cred directly calculated by the CredRank algorithm.
+ */
+export type OutputNode = {|
+  +address: $ReadOnlyArray<string>,
+  +cred: number,
+  +mint: number,
+  +timestamp: TimestampMs | null,
+  +description: string,
+|};
+
+/**
+ * The first version of the output format will have the nodes, and the plugin
+ * declarations. Including the declarations makes it easy for consumers to do
+ * analyses like: how much cred was minted within each plugin? Within each
+ * contribution type? Etc.
+ */
+export type OutputV1 = {|
+  // Ordered by address
+  +orderdNodes: $ReadOnlyArray<OutputNode>,
+  +plugins: $ReadOnlyArray<PluginDeclaration>,
+|};
+
+/**
+ * Extra data for Contributors. Note that each Contributor corresponds
+ * to a specific node in the graph, which is retrievable via the NodeIndex.
+ *
+ * For contributors, we'll give the cred-over-time. We don't include this
+ * information for regular contribution nodes, both because it's not that
+ * interesting, and because once we migrate to CredRank, we won't have that
+ * level of resolution anymore.
+ *
+ * While we are in TimelineCred, the credByTime will be the raw cred that the
+ * contributor recieved in each time slice. Once we switch to TimelineCred, the
+ * credByTime will instead be the amount of cred that flowed from the
+ * contributor's epoch node for each time period.
+ *
+ * In either case, it will be the an invariant that the sum of the credByTime
+ * will correspond to the cred score on the raw OutputNode.
+ */
+export type Contributor = {|
+  +alias: Alias,
+  // Indexed into the orderedNodes
+  +nodeIndex: Index,
+  // Matches the interval endpoints in the full output format.
+  +credByTime: $ReadOnlyArray<number>,
+|};
+
+/**
+ * The second version of the output format includes explicit contributor data,
+ * with cred-over-time and convenient aliases for looking up particular
+ * contributors.
+ */
+export type OutputV2 = {|
+  +orderdNodes: $ReadOnlyArray<OutputNode>,
+  +plugins: $ReadOnlyArray<PluginDeclaration>,
+  +contributors: $ReadOnlyArray<Contributor>,
+  +intervalEnd: $ReadOnlyArray<TimestampMs>,
+|};
+
+/**
+ * This edge format includes the standard Graph.Edge information,
+ * along with the edge weight and (most interestingly), the
+ * cred flow across the edge. In TimelineCred, the cred flow
+ * will be summed across every time period; once we switch to CredRank,
+ * the summing will no longer be necessary.
+ */
+export type OutputEdge = {|
+  +address: $ReadOnlyArray<string>,
+  +edgeWeight: EdgeWeight,
+  +credFlow: EdgeWeight,
```

Comment:
> Maybe it would be better if in addition to including the raw
> EdgeWeights, we also have the normalized transitionProbabilities, in
> which case we could give webbing edges or sponsorship edges or boost
> edges a null EdgeWeight, but they would still have the
> transitionProbabilties?

My reading of this line (`credFlow: EdgeWeight`) was that this *was*
what was meant, and that the `EdgeWeight` type was being kind of coöpted
as a generic “bidirectional float”. If that’s not what was meant, please
clarify.

Reply:
Yeah, I co-opted the EdgeWeight as a sort of bidirectional float. My intention was for `EdgeWeight` to contain the original (unnormalized) edge weights, but I think that having it be the transition probability could be cleaner.

CredFlow shows how much cred flowed in each direction--re using the EdgeWeight type may be confusing. My intention at the moment is that the `edgeWeight` field will have the raw (unnormalized) edge weights, but I think we can take it out for now. In TimelineCred it's not tremendously helpful, and in CredRank we may want transition probabilties instead.